Lushu/src/main/kotlin/lushu/ContextGrammar/Grammar/Rules.kt

```diff
@@ -0,0 +1,250 @@
+                val matcher = pattern.matcher(word)
+                if (matcher.find()) index else null
+            }
+        }
+        return matchingIndex.distinct().sorted()
+    }
+
+    /**
+     * Converts a list of tokens into cipher tokens.
+     * This function iterates over the input list and checks if it contains any text patterns that need to be encrypted.
+     * The resulting cipher tokens are returned as a list.
+     * @param inputTokens The list of input tokens to be converted into cipher tokens.
+     * @return The list of cipher tokens generated from the input tokens.
+     * Example:
+     * Let a context "host ip: 123.123.123.123" where 123.123.123.123 needs to be encrypted.
+     * Input: ["server", "host", "ip:",  "123.123.123.123", "is", "online"]
+     * Output: ["server", "host", "ip:",  "***************", "is", "online"]
+     */
+    fun tokens2CipherTokens(inputTokens: List<String>): List<String> {
+        val regexList: List<String> = root.getChildren().map { child -> child.getRegex() }
+        val matchingIndex: List<Int> = findMatchingIndex(inputTokens, regexList)
+
+        var i = defaultMinimalIndex
+
+        val cipherTokens = inputTokens.toMutableList()
+
+        matchingIndex.forEach { index ->
+            if (i <= index) {
+                i = matchTokensAgainstPatternContext(cipherTokens, inputTokens.toMutableList(), index, root)
+            }
+        }
+
+        return cipherTokens
+    }
+
+    /**
+     * Adds rules based on the given context rule by recursively processing each word in the context rule.
+     * @param contextRule The mutable list of strings representing the context rule.
+     * @param current The current node in the tree structure. Defaults to the root node.
+     *
+     * - Example:
+     * Input: mutableListOf("This", "<t>is", "an</t>", "<s>example</s>", "sentence.")
+     */
+    private fun addContextToGrammar(contextRule: MutableList<String>, current: Node? = root) {
+        if (contextRule.isNullOrEmpty()) {
+            return
+        }
+        val firstWord = 0
+
+        val (isCase, nextCase) = dsl.hasTags(contextRule[firstWord])
+
+        val word = dsl.removeAllTagsFromWord(contextRule[firstWord])
+
+        dsl.setIsCase(isCase)
+
+        val updatedCurrent = current?.addChild(
+            word,
+            s = dsl.isSensitive(),
+            pc = dsl.isPlus(),
+            npc = nextCase[isStar],
+            m = dsl.isTerminal(),
+        )
+
+        dsl.setIsCase(nextCase)
+
+        contextRule.removeAt(firstWord)
+        addContextToGrammar(contextRule, updatedCurrent)
+    }
+
+    /**
+     * Retrieves the context from the given words and adds rules based on the context.
+     * @param words The string containing the words from which the context is extracted.
+     *
+     * - Example:
+     * Input: This is an example sentence <c>with context</c> and <c>another context</c>
+     */
+    fun getContext(words: String?) {
```

Comment:
```suggestion
    fun addContextsFromWords(words: String?) {
```
`get` functions in OOP usually return something. Please rename this function to something similar to the above